File: src/main/java/hiiragi283/material/api/material/MaterialStack.java
package hiiragi283.material.api.material;

import hiiragi283.material.util.HiiragiUtil;
import net.minecraft.nbt.NBTTagCompound;
import net.minecraftforge.fluids.Fluid;
import net.minecraftforge.fluids.FluidStack;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.List;
import java.util.Objects;
import java.util.Optional;


public class MaterialStack {

    @Nullable
    private final HiiragiMaterial material;
    public int amount;

    public MaterialStack(@Nullable HiiragiMaterial material, int amount) {
        this.material = material;
        this.amount = amount;
    }

    public MaterialStack(FluidStack fluidStack) {
        this(fluidStack.getFluid(), fluidStack.amount);
    }

    public MaterialStack(Fluid fluid, int amount) {
        this(fluid.getName(), amount);
    }

    public MaterialStack(NBTTagCompound nbt) {
        this(nbt.getString(HiiragiUtil.MATERIAL), nbt.getInteger(HiiragiUtil.AMOUNT));
    }

    public MaterialStack(String name, int amount) {
        this(HiiragiMaterial.REGISTRY.getValue(name).orElse(null), amount);
    }

    public MaterialStack(MaterialStack stack, int amount) {
        this(stack.material, amount);
    }

    public Optional<HiiragiMaterial> getMaterial() {
        return Optional.ofNullable(material);
    }

    public static MaterialStack EMPTY = new MaterialStack((HiiragiMaterial) null, 0);

    public void addTooltip(List<String> tooltip) {
        if (material == null) return;
        material.addTooltip(tooltip, material.getTranslatedName(), amount);
    }

    public String getTranslatedName() {
        return material != null ? material.getTranslatedName() : "";
    }

    public boolean isEmpty() {
        return Objects.equals(this, EMPTY) || this.material == null || this.amount <= 0;
    }

    //    Amount    //

    public MaterialStack addAmount(int add) {
        return new MaterialStack(this.material, this.amount + add);
    }

    public MaterialStack removeAmount(int remove) {
        return new MaterialStack(this.material, this.amount - remove);
    }

    public MaterialStack setAmount(int amount) {
        return new MaterialStack(this.material, amount);
    }

    public boolean equalsMaterial(@NotNull MaterialStack other) {
        return material != null && this.material.equals(other.material);
    }

    //    Serialization    //

    public NBTTagCompound serializeNBT() {
        var tag = new NBTTagCompound();
        tag.setString(HiiragiUtil.MATERIAL, material != null ? material.name() : "");
        tag.setInteger(HiiragiUtil.AMOUNT, amount);
        return tag;
    }

}

File: src/main/kotlin/hiiragi283/material/api/material/MaterialStack.kt
package hiiragi283.material.api.material

import hiiragi283.material.api.registry.HiiragiRegistries
import net.minecraft.nbt.NBTTagCompound
import net.minecraftforge.event.entity.player.ItemTooltipEvent
import net.minecraftforge.fluids.Fluid
import net.minecraftforge.fluids.FluidStack

data class MaterialStack(val material: HiiragiMaterial, var amount: Int) {

    companion object {

        @JvmStatic
        fun of(compound: NBTTagCompound) = of(compound.getString("material"), compound.getInteger("amount"))

        @JvmStatic
        fun of(fluid: Fluid, amount: Int) = of(fluid.name, amount)

        @JvmStatic
        fun of(fluidStack: FluidStack) = of(fluidStack.fluid, fluidStack.amount)

        @JvmStatic
        fun of(name: String, amount: Int): MaterialStack? =
            HiiragiRegistries.MATERIAL.getValue(name)?.let { MaterialStack(it, amount) }

    }

    fun addTooltip(event: ItemTooltipEvent) {
        addTooltip(event.toolTip)
    }

    fun addTooltip(tooltip: MutableList<String>) {
        material.addTooltip(tooltip, material.getTranslatedName(), amount)
    }

    //    Amount    //

    fun equalsMaterial(other: MaterialStack): Boolean = this.material == other.material

    //    Serialization    //

    fun serializeNBT(): NBTTagCompound = NBTTagCompound().also {
        it.setString("material", material.name)
        it.setInteger("amount", amount)
    }

    //    General    //

    override fun toString(): String = "$material:$amount"

}